
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="test site">
      
      
        <meta name="author" content="Princess Rodriguez">
      
      
        <link rel="canonical" href="https://prodriguez19.github.io/MMG3320-5320/ch09/17_counting_reads/">
      
      
        <link rel="prev" href="../../ch08/16_RSeQC/">
      
      
        <link rel="next" href="../../assignments/week1_assignment/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.42">
    
    
      
        <title>HTSeq (L17) - MMG:3320/5320 Advanced Bioinformatics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.0253249f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-objectives" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="MMG:3320/5320 Advanced Bioinformatics" class="md-header__button md-logo" aria-label="MMG:3320/5320 Advanced Bioinformatics" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MMG:3320/5320 Advanced Bioinformatics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              HTSeq (L17)
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../slides/" class="md-tabs__link">
          
  
  Slides

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../pdf_materials/" class="md-tabs__link">
          
  
  Printouts

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../ch01/" class="md-tabs__link">
          
  
  Lessons

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../assignments/week1_assignment/" class="md-tabs__link">
          
  
  Assignments

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="MMG:3320/5320 Advanced Bioinformatics" class="md-nav__button md-logo" aria-label="MMG:3320/5320 Advanced Bioinformatics" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    MMG:3320/5320 Advanced Bioinformatics
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Home
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../MMG3320-Spring-2025-Syllabus.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Syllabus
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../MMG3320-5320-Spring-2025-Schedule.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/gitbash-install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Git Bash Install
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../FAQ/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Frequently Asked Questions
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/quick-start/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Slides
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Slides
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../slides/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Printouts
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Printouts
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pdf_materials/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Lessons
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Lessons
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 1
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Week 1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/01_week1_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introducing the Shell (L1)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/02_week1_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Navigating the Shell (L2)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Week 2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch02/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch02/03_week2_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Wild Cards & Nano (L3)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 3
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Week 3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch03/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch03/04_week3_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Searching and Redirection (L4)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch03/05_week3_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intro to Shell Scripting (L5)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 4
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            Week 4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch04/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch04/06_week4_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Loops (L6)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch04/07_week4_intro-to-RNAseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA-Seq Overview (L7)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch04/08_accessing_public_experimental_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Accessing Public Data (L8)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_5" >
        
          
          <label class="md-nav__link" for="__nav_4_5" id="__nav_4_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 5
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_5">
            <span class="md-nav__icon md-icon"></span>
            Week 5
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch05/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch05/09_assessing_fastqc_output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC with FASTQC (L9)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch05/10_trimming/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Trimming and Filtering (L10)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_6" >
        
          
          <label class="md-nav__link" for="__nav_4_6" id="__nav_4_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 6
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_6">
            <span class="md-nav__icon md-icon"></span>
            Week 6
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch06/11_ClassExercise_Align-Submission/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Submitting Alignment Script (L11)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch06/13_ClassExercise3_Index_Genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Indexing a Reference Genome (L12)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch06/12_ClassExercise2_Align-Modify/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modifying Alignment Script (L13)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_7" >
        
          
          <label class="md-nav__link" for="__nav_4_7" id="__nav_4_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 7
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_7">
            <span class="md-nav__icon md-icon"></span>
            Week 7
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch07/14_instructions-for-course-catchup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course Catch Up Day (CCU)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch07/14_installing-RSeQC-using-conda/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installing RSeQC (L14)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_8" >
        
          
          <label class="md-nav__link" for="__nav_4_8" id="__nav_4_8_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 8
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_8">
            <span class="md-nav__icon md-icon"></span>
            Week 8
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch08/16_RSeQC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RSeQC (L16)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_9" checked>
        
          
          <label class="md-nav__link" for="__nav_4_9" id="__nav_4_9_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 9
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_9_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_9">
            <span class="md-nav__icon md-icon"></span>
            Week 9
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    HTSeq (L17)
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    HTSeq (L17)
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Learning Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#counting-reads-as-a-measure-of-gene-expression" class="md-nav__link">
    <span class="md-ellipsis">
      Counting reads as a measure of gene expression
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Workflow
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Basic Workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input Files:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#counting-reads" class="md-nav__link">
    <span class="md-ellipsis">
      Counting Reads:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#counting-reads-in-features-with-htseq-count" class="md-nav__link">
    <span class="md-ellipsis">
      Counting reads in features with htseq-count
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-htseq-count" class="md-nav__link">
    <span class="md-ellipsis">
      Running htseq-count
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Running htseq-count">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-command-syntax" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Command Syntax
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Basic Command Syntax">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#options" class="md-nav__link">
    <span class="md-ellipsis">
      Options
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-considerations" class="md-nav__link">
    <span class="md-ellipsis">
      Key Considerations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#limitations-alternatives" class="md-nav__link">
    <span class="md-ellipsis">
      Limitations &amp; Alternatives
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Limitations & Alternatives">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#htseq-count-output" class="md-nav__link">
    <span class="md-ellipsis">
      htseq-count output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resource-recommendations-for-htseq-count" class="md-nav__link">
    <span class="md-ellipsis">
      Resource Recommendations for htseq-count
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#creating-bigwig-files" class="md-nav__link">
    <span class="md-ellipsis">
      Creating bigWig files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Assignments
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Assignments
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week1_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 1 (L2)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week2_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 2 (L3)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week3_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 3 (L5)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week4_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 4 (L6)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week5_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 5
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week6_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 6
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Final Project
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/assignment_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assignment Schedule
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Learning Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#counting-reads-as-a-measure-of-gene-expression" class="md-nav__link">
    <span class="md-ellipsis">
      Counting reads as a measure of gene expression
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Workflow
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Basic Workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input Files:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#counting-reads" class="md-nav__link">
    <span class="md-ellipsis">
      Counting Reads:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#counting-reads-in-features-with-htseq-count" class="md-nav__link">
    <span class="md-ellipsis">
      Counting reads in features with htseq-count
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-htseq-count" class="md-nav__link">
    <span class="md-ellipsis">
      Running htseq-count
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Running htseq-count">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-command-syntax" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Command Syntax
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Basic Command Syntax">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#options" class="md-nav__link">
    <span class="md-ellipsis">
      Options
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-considerations" class="md-nav__link">
    <span class="md-ellipsis">
      Key Considerations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#limitations-alternatives" class="md-nav__link">
    <span class="md-ellipsis">
      Limitations &amp; Alternatives
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Limitations & Alternatives">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#htseq-count-output" class="md-nav__link">
    <span class="md-ellipsis">
      htseq-count output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resource-recommendations-for-htseq-count" class="md-nav__link">
    <span class="md-ellipsis">
      Resource Recommendations for htseq-count
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#creating-bigwig-files" class="md-nav__link">
    <span class="md-ellipsis">
      Creating bigWig files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>HTSeq (L17)</h1>

<div><h2 id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Conversion of BAM file to a counts file using <code>htseq-count</code></li>
<li>Conversion of BAM file to a bigWig file using <code>deeptools</code> </li>
</ul>
<h2 id="counting-reads-as-a-measure-of-gene-expression">Counting reads as a measure of gene expression</h2>
<p>Once we have our reads aligned to the genome, the next step is to count how many reads have mapped to each gene. Many tools can take BAM files as input and output the number of reads (counts) associated with each feature of interest (genes, exons, transcripts, etc.). Two commonly used counting tools are featureCounts and <a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html">htseq-count</a>.</p>
<figure>
  <img alt="shells" src="../../img/counts-workflow.png" width="400">
</figure>
<ul>
<li>
<p>These tools report raw counts, meaning they only count reads that map uniquely to a single location in the genome. They are most effective for counting at the gene level. In this approach, the total read count for a gene (meta-feature) is the sum of reads assigned to each of its exons (features).</p>
</li>
<li>
<p>Some other tools account for multiple transcripts per gene, assigning fractional counts instead of whole numbers. For example, if one read aligns to two transcripts, it may be counted as 0.5 for each transcript rather than a whole number.</p>
</li>
<li>
<p>There are also tools that count multimapping reads, but this approach can lead to overcounting, which affects normalization and ultimately compromises the accuracy of differential gene expression analysis.</p>
</li>
</ul>
<p><strong>Input for counting = multiple BAM files + 1 GTF file</strong></p>
<p>Simply speaking, the genomic coordinates of where the read is mapped (BAM) are cross-referenced with the genomic coordinates of whichever feature you are interested in counting expression of (GTF), it can be exons, genes or transcripts.</p>
<figure>
  <img alt="shells" src="../../img/count-fig2.png" width="600">
</figure>
<p><strong>Output of counting = A count matrix, with genes as rows and samples are columns</strong></p>
<p>These are the "raw" counts and will be used in statistical programs downstream for differential gene expression. Below is a representative counts matrix. </p>
<figure>
  <img alt="shells" src="../../img/count-matrix.png" width="400">
</figure>
<p>We will begin by demonstrating the usage of HTSeq. HTSeq-count is a command-line tool used in RNA-Seq data analysis to count the number of sequencing reads that align to specific genomic features, such as genes. It is part of the HTSeq Python package and is commonly used in differential gene expression analysis pipelines.</p>
<h2 id="basic-workflow">Basic Workflow</h2>
<h3 id="input-files">Input Files:</h3>
<ul>
<li>SAM/BAM file: Contains aligned RNA-Seq reads (output from tools like HISAT2 or STAR).</li>
<li>GTF/GFF file: Contains gene annotations specifying genomic coordinates.</li>
</ul>
<h3 id="counting-reads">Counting Reads:</h3>
<ul>
<li>HTSeq-count assigns each read to a genomic feature (usually a gene) based on predefined rules.</li>
<li>It outputs a table listing each gene and the number of reads assigned to it.</li>
</ul>
<h3 id="output">Output:</h3>
<ul>
<li>A tab-delimited text file with two columns:<ul>
<li>Gene ID</li>
<li>Read count</li>
</ul>
</li>
</ul>
<h2 id="counting-reads-in-features-with-htseq-count">Counting reads in features with <code>htseq-count</code></h2>
<p>Given a file with aligned sequencing reads and a list of genomic features, a common task is to count how many reads map to each feature.</p>
<p>Here, a feature refers to a specific interval (i.e., a range of positions) on a chromosome or a union of such intervals.</p>
<p>Since our example data comes from an RNA-Seq experiment, we aim to count how many reads fall within the exonic regions of each gene. To do this, we first need information about exon positions, which can be obtained from GTF files—a common format provided by Ensembl (available here).</p>
<p>Special care must be taken when handling reads that align to or overlap with multiple features. The htseq-count script offers <strong>three different modes</strong> to handle such cases.</p>
<ol>
<li>
<p><strong>Union (Recommended for most cases):</strong></p>
<ul>
<li>A read is assigned to a feature if any part of the read overlaps with it.</li>
<li>If a read overlaps multiple features, it is not counted at all (to avoid ambiguity).</li>
</ul>
</li>
<li>
<p><strong>Intersection-strict:</strong></p>
<ul>
<li>A read is assigned to a feature only if every position of the read overlaps with that feature.</li>
<li>If a read overlaps multiple features but not completely within one, it is not counted.</li>
</ul>
</li>
<li>
<p><strong>Intersection-nonempty:</strong></p>
<ul>
<li>A read is assigned to a feature only if it overlaps with at least one feature at every position.</li>
<li>Unlike intersection-strict, it ignores positions where no features are present.</li>
</ul>
</li>
</ol>
<p>The figure below illustrates the effect of these three modes: </p>
<figure>
  <img alt="shells" src="../../img/count_modes.png" width="400">
</figure>
<h2 id="running-htseq-count">Running <code>htseq-count</code></h2>
<p>Begin by loading the <code>htseq-count</code> module with <code>module load</code></p>
<div class="highlight"><pre><span></span><code>module<span class="w"> </span>load<span class="w"> </span>gcc/13.3.0-xp3epyt<span class="w"> </span>
py-htseq/2.0.3-mb7ap7s
</code></pre></div>
<p>Run the following to check the module is loaded: </p>
<div class="highlight"><pre><span></span><code>htseq-count<span class="w"> </span>--help
</code></pre></div>
<div class="highlight"><pre><span></span><code>usage: htseq-count [-h] [--version] [-f {sam,bam,auto}] [-r {pos,name}] [--max-reads-in-buffer MAX_BUFFER_SIZE] [-s {yes,no,reverse}] [-a MINAQUAL]
                   [-t FEATURE_TYPE] [-i IDATTR] [--additional-attr ADDITIONAL_ATTRIBUTES] [--add-chromosome-info]
                   [-m {union,intersection-strict,intersection-nonempty}] [--nonunique {none,all,fraction,random}]
                   [--secondary-alignments {score,ignore}] [--supplementary-alignments {score,ignore}] [-o SAMOUTS] [-p {SAM,BAM,sam,bam}]
                   [-d OUTPUT_DELIMITER] [-c OUTPUT_FILENAME] [--counts-output-sparse] [--append-output] [-n NPROCESSES] [--feature-query FEATURE_QUERY]
                   [-q] [--with-header]
                   samfilenames [samfilenames ...] featuresfilename
This script takes one or more alignment files in SAM/BAM format and a feature file in GFF format and calculates for each feature the number of reads
mapping to it. See http://htseq.readthedocs.io/en/master/count.html for details.
</code></pre></div>
<h3 id="basic-command-syntax">Basic Command Syntax</h3>
<div class="highlight"><pre><span></span><code>htseq-count<span class="w"> </span>-f<span class="w"> </span>bam<span class="w"> </span>-s<span class="w"> </span>no<span class="w"> </span>-i<span class="w"> </span>gene_id<span class="w"> </span>sample1.bam<span class="w"> </span>genes.gtf<span class="w"> </span>&gt;<span class="w"> </span>gene_counts.txt
</code></pre></div>
<ul>
<li><code>-f bam</code> → Specifies input format (can be <code>sam</code> or <code>bam</code>).</li>
<li><code>-s no</code> → Defines strandedness (<code>yes</code>, <code>no</code>, or <code>reverse</code>).</li>
<li><code>-i gene_id</code> → used to identify the GTF feature attribute (<code>gene_name</code> or <code>gene_id</code>)</li>
<li><code>aligned_reads.bam</code> → Input file with mapped reads.</li>
<li><code>genes.gtf</code> → Reference annotation file.</li>
<li><code>&gt; gene_counts.txt</code> → Redirects output to a file.</li>
</ul>
<h4 id="options">Options</h4>
<div class="highlight"><pre><span></span><code>-f<span class="w"> </span>&lt;format&gt;,<span class="w"> </span>--format<span class="o">=</span>&lt;format&gt;

<span class="w">    </span>Format<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>input<span class="w"> </span>data.<span class="w"> </span>Possible<span class="w"> </span>values<span class="w"> </span>are<span class="w"> </span>sam<span class="w"> </span><span class="o">(</span><span class="k">for</span><span class="w"> </span>text<span class="w"> </span>SAM<span class="w"> </span>files<span class="o">)</span><span class="w"> </span>and<span class="w"> </span>bam<span class="w"> </span><span class="o">(</span><span class="k">for</span><span class="w"> </span>binary<span class="w"> </span>BAM<span class="w"> </span>files<span class="o">)</span>.<span class="w"> </span>Default<span class="w"> </span>is<span class="w"> </span>sam.
</code></pre></div>
<div class="highlight"><pre><span></span><code>-s<span class="w"> </span>&lt;yes/no/reverse&gt;,<span class="w"> </span>--stranded<span class="o">=</span>&lt;yes/no/reverse&gt;

<span class="w">    </span>whether<span class="w"> </span>the<span class="w"> </span>data<span class="w"> </span>is<span class="w"> </span>from<span class="w"> </span>a<span class="w"> </span>strand-specific<span class="w"> </span>assay<span class="w"> </span><span class="o">(</span>default:<span class="w"> </span><span class="sb">`</span>yes<span class="sb">`</span><span class="o">)</span>

<span class="w">    </span>For<span class="w"> </span><span class="nv">stranded</span><span class="o">=</span>no,<span class="w"> </span>a<span class="w"> </span><span class="nb">read</span><span class="w"> </span>is<span class="w"> </span>considered<span class="w"> </span>overlapping<span class="w"> </span>with<span class="w"> </span>a<span class="w"> </span>feature<span class="w"> </span>regardless<span class="w"> </span>of<span class="w"> </span>whether<span class="w"> </span>it<span class="w"> </span>is<span class="w"> </span>mapped<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>same<span class="w"> </span>or<span class="w"> </span>the<span class="w"> </span>opposite<span class="w"> </span>strand<span class="w"> </span>as<span class="w"> </span>the<span class="w"> </span>feature.<span class="w"> </span>For<span class="w"> </span><span class="nv">stranded</span><span class="o">=</span>yes<span class="w"> </span>and<span class="w"> </span>single-end<span class="w"> </span>reads,<span class="w"> </span>the<span class="w"> </span><span class="nb">read</span><span class="w"> </span>has<span class="w"> </span>to<span class="w"> </span>be<span class="w"> </span>mapped<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>same<span class="w"> </span>strand<span class="w"> </span>as<span class="w"> </span>the<span class="w"> </span>feature.<span class="w"> </span>For<span class="w"> </span>paired-end<span class="w"> </span>reads,<span class="w"> </span>the<span class="w"> </span>first<span class="w"> </span><span class="nb">read</span><span class="w"> </span>has<span class="w"> </span>to<span class="w"> </span>be<span class="w"> </span>on<span class="w"> </span>the<span class="w"> </span>same<span class="w"> </span>strand<span class="w"> </span>and<span class="w"> </span>the<span class="w"> </span>second<span class="w"> </span><span class="nb">read</span><span class="w"> </span>on<span class="w"> </span>the<span class="w"> </span>opposite<span class="w"> </span>strand.<span class="w"> </span>For<span class="w"> </span><span class="nv">stranded</span><span class="o">=</span>reverse,<span class="w"> </span>these<span class="w"> </span>rules<span class="w"> </span>are<span class="w"> </span>reversed.
</code></pre></div>
<div class="highlight"><pre><span></span><code>-i<span class="w"> </span>&lt;id<span class="w"> </span>attribute&gt;,<span class="w"> </span>--idattr<span class="o">=</span>&lt;id<span class="w"> </span>attribute&gt;

<span class="w">    </span>GFF<span class="w"> </span>attribute<span class="w"> </span>to<span class="w"> </span>be<span class="w"> </span>used<span class="w"> </span>as<span class="w"> </span>feature<span class="w"> </span>ID.<span class="w"> </span>Several<span class="w"> </span>GFF<span class="w"> </span>lines<span class="w"> </span>with<span class="w"> </span>the<span class="w"> </span>same<span class="w"> </span>feature<span class="w"> </span>ID<span class="w"> </span>will<span class="w"> </span>be<span class="w"> </span>considered<span class="w"> </span>as<span class="w"> </span>parts<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>same<span class="w"> </span>feature.<span class="w"> </span>The<span class="w"> </span>feature<span class="w"> </span>ID<span class="w"> </span>is<span class="w"> </span>used<span class="w"> </span>to<span class="w"> </span>identity<span class="w"> </span>the<span class="w"> </span>counts<span class="w"> </span><span class="k">in</span><span class="w"> </span>the<span class="w"> </span>output<span class="w"> </span>table.<span class="w"> </span>The<span class="w"> </span>default,<span class="w"> </span>suitable<span class="w"> </span><span class="k">for</span><span class="w"> </span>RNA-Seq<span class="w"> </span>analysis<span class="w"> </span>using<span class="w"> </span>an<span class="w"> </span>Ensembl<span class="w"> </span>GTF<span class="w"> </span>file,<span class="w"> </span>is<span class="w"> </span><span class="sb">`</span>gene_id<span class="sb">`</span>.
</code></pre></div>
<h2 id="key-considerations">Key Considerations</h2>
<p><strong>Stranded vs. Unstranded Data:</strong>
  + Many RNA-Seq protocols preserve strand information. Setting -s yes or -s reverse ensures proper assignment.</p>
<p><strong>Feature Type:</strong>
  + HTSeq-count assigns reads based on featuretype (default is exon in GTF). If needed, use -t to specify other features.</p>
<p><strong>Overlap Mode:</strong>
  + Some reads may overlap multiple features; the -m flag determines how they are assigned (union, intersection-strict, intersection-nonempty).</p>
<h2 id="limitations-alternatives">Limitations &amp; Alternatives</h2>
<ul>
<li>HTSeq-count is a gene-level summarization tool; it does not handle transcript-level quantification.</li>
<li>It requires sorted BAM/SAM files.</li>
<li>Alternative tools like featureCounts (..which is faster) or Salmon/RSEM (for transcript-level quantification) may be preferred in some cases.</li>
</ul>
<hr>
<div class="admonition example">
<p class="admonition-title">Class Exercise: HTSeq-count Instructions and Script</p>
<ol>
<li><strong>Copy Folder:</strong> Make a copy of the following folder into your home directory: </li>
<li><strong>Determine Strandedness:</strong> HtSeq-count requires setting the <code>-s</code> parameter based on the RNA-Seq library preparation protocol. The options are:<ul>
<li><code>-s yes</code>, reads are mapped to the same strand as the sense strand </li>
<li><code>-s no</code>, reads can map to either strand (unstranded)</li>
<li><code>-s reverse</code>, reads are mapped to the opposite strand (anti-sense)</li>
</ul>
</li>
<li><strong>Generate and submit script:</strong> You will need to add/specify/modify the following parameters:</li>
<li>add the module </li>
<li>add the path to the GTF file: </li>
<li><code>-s</code> options include yes, no, or reverse</li>
<li><code>-i</code> specify <code>gene_id</code></li>
<li>Submit the script when you are done using the example script below. </li>
<li>Rerun the script, this time change to <code>-i gene_name</code> instead. Be sure to change the name of the output file. What is the difference in the outputs? </li>
</ol>
</div>
<p>The example script below will get you started: </p>
<div class="highlight"><pre><span></span><code><span class="ch">#!/bin/bash</span>
<span class="c1">#SBATCH --partition=general </span>
<span class="c1">#SBATCH --nodes=1</span>
<span class="c1">#SBATCH --ntasks=4</span>
<span class="c1">#SBATCH --mem=10G</span>
<span class="c1">#SBATCH --time=30:00:00</span>
<span class="c1">#SBATCH --job-name=htseq-count    </span>
<span class="c1">#SBATCH --output=%x_%j.out  # %x=job-name, %j=jobid</span>

<span class="c1"># Load HTSeq module</span>


<span class="c1"># Define the path to the GTF file</span>
<span class="nv">GTF_FILE</span><span class="o">=</span>

<span class="c1"># Iterate through all BAM files in the current directory</span>
<span class="k">for</span><span class="w"> </span>BAM_FILE<span class="w"> </span><span class="k">in</span><span class="w"> </span>*.bam<span class="p">;</span><span class="w"> </span><span class="k">do</span>

<span class="w">  </span><span class="c1"># Extract the filename without the .bam extension</span>
<span class="w">  </span><span class="nv">NAME</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="s2">"</span><span class="nv">$BAM_FILE</span><span class="s2">"</span><span class="w"> </span>.bam<span class="k">)</span>
<span class="w">  </span><span class="nb">echo</span><span class="w"> </span><span class="s2">"Processing: </span><span class="nv">$NAME</span><span class="s2">"</span>

<span class="w">  </span><span class="c1"># Run HTSeq-count</span>
<span class="w">  </span>htseq-count<span class="w"> </span>-f<span class="w"> </span>bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-s<span class="w"> </span><span class="se">\ </span><span class="w"> </span><span class="c1"># Modify based on your library strandedness</span>
<span class="w">    </span>-i<span class="w"> </span>gene_id<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-m<span class="w"> </span>union<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="s2">"</span><span class="nv">$BAM_FILE</span><span class="s2">"</span><span class="w"> </span><span class="s2">"</span><span class="nv">$GTF_FILE</span><span class="s2">"</span><span class="w"> </span>&gt;<span class="w"> </span><span class="s2">"</span><span class="si">${</span><span class="nv">NAME</span><span class="si">}</span><span class="s2">.gene_id.count"</span>

<span class="k">done</span>
</code></pre></div>
<h3 id="htseq-count-output"><code>htseq-count</code> output</h3>
<p>The output of this tool is 2 files, <em>a count matrix</em> and <em>a summary file</em> that tabulates how many the reads were "assigned" or counted and the reason they remained "unassigned". Let's take a look at the summary file:</p>
<p></p><div class="highlight"><pre><span></span><code>$<span class="w"> </span>less<span class="w"> </span>results/counts/Mov10_featurecounts.txt.summary
</code></pre></div>
Now let's look at the count matrix:
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>less<span class="w"> </span>results/counts/Mov10_featurecounts.txt
</code></pre></div>
<h2 id="resource-recommendations-for-htseq-count">Resource Recommendations for <code>htseq-count</code></h2>
<p>An array will submit independent jobs and process each BAM file independently. </p>
<div class="highlight"><pre><span></span><code><span class="ch">#!/bin/bash</span>
<span class="c1">#SBATCH --partition=general</span>
<span class="c1">#SBATCH --array=1-8  # Adjust for the number of samples (1 per sample)</span>
<span class="c1">#SBATCH --nodes=1</span>
<span class="c1">#SBATCH --ntasks=1  </span>
<span class="c1">#SBATCH --mem=4G  # 2-6GB, 4GB is typically enough but specify 6GB for 100M+ reads</span>
<span class="c1">#SBATCH --time=2:00:00</span>
<span class="c1">#SBATCH --job-name=htseq-array</span>
<span class="c1">#SBATCH --output=htseq-%A_%a.out  # %A = job ID, %a = array task ID</span>
</code></pre></div>
<h2 id="creating-bigwig-files">Creating bigWig files</h2>
<p>We will now take our BAM files and convert them into bigWig files. The bigWig format is an indexed binary format useful for dense, continuous data that will be displayed in a genome browser as a graph/track, but also is used as input for some of the visualization commands we will be running in <code>deepTools</code>. </p>
<p><a href="http://deeptools.readthedocs.org/en/latest/content/list_of_tools.html"><code>deepTools</code></a>, is a suite of Python tools developed for the efficient analysis of high-throughput sequencing data, such as ChIP-seq, RNA-seq or MNase-seq. <code>deepTools</code> has a wide variety of commands that go beyond those that are covered in this lesson. We encourage you to look through the documentation and explore on your own time. </p>
<figure>
  <img alt="shells" src="../../img/bam_to_bigwig.png" width="700">
</figure>
<p><em>Image acquired from <a href="http://deeptools.readthedocs.io/en/latest/content/tools/bamCoverage.html?highlight=bigwig">deepTools documentation</a> pages</em></p>
<hr>
<p><em>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</em></p></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["content.code.copy", "content.code.annotate", "content.code.select", "navigation.tabs", "navigation.top"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.83f73b43.min.js"></script>
      
    
  </body>
</html>