
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="test site">
      
      
        <meta name="author" content="Princess Rodriguez">
      
      
        <link rel="canonical" href="https://prodriguez19.github.io/MMG3320-5320/ch04/07_week4_intro-to-RNAseq/">
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.42">
    
    
      
        <title>07 week4 intro to RNAseq - MMG:3320 Advanced Bioinformatics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.0253249f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-objectives" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="MMG:3320 Advanced Bioinformatics" class="md-header__button md-logo" aria-label="MMG:3320 Advanced Bioinformatics" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MMG:3320 Advanced Bioinformatics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              07 week4 intro to RNAseq
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../slides/" class="md-tabs__link">
          
  
  Slides

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../pdf_materials/" class="md-tabs__link">
          
  
  Printouts

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../ch01/" class="md-tabs__link">
          
  
  Lessons

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../assignments/week1_assignment/" class="md-tabs__link">
          
  
  Assignments

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="MMG:3320 Advanced Bioinformatics" class="md-nav__button md-logo" aria-label="MMG:3320 Advanced Bioinformatics" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    MMG:3320 Advanced Bioinformatics
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Home
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pdf_materials/MMG3320-Spring-2026-Syllabus.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Syllabus
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pdf_materials/MMG3320-Spring-2026-Schedule.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/gitbash-install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Git Bash Install
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../FAQ/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Frequently Asked Questions
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/quick-start/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Slides
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Slides
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../slides/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Printouts
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Printouts
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pdf_materials/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Lessons
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Lessons
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 1
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Week 1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/01_week1_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introducing the Shell (L1)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch01/02_week1_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Navigating the Shell (L2)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Week 2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Week 2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch02/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ch02/03_week2_mmg3320/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Wild Cards & Nano (L3)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Assignments
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Assignments
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/week1_assignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homework 1 (L2)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assignments/assignment_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assignment Schedule
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Learning Objectives:
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#overview-of-rna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      Overview of RNA-seq
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview of RNA-seq">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#transcriptomics" class="md-nav__link">
    <span class="md-ellipsis">
      Transcriptomics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#illumina-sequencing" class="md-nav__link">
    <span class="md-ellipsis">
      Illumina Sequencing
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Illumina Sequencing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#illumina-library-preparation" class="md-nav__link">
    <span class="md-ellipsis">
      Illumina Library preparation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#strandedness" class="md-nav__link">
    <span class="md-ellipsis">
      Strandedness
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#single-end-versus-paired-end" class="md-nav__link">
    <span class="md-ellipsis">
      Single-end versus Paired-end
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#different-sequencing-platforms" class="md-nav__link">
    <span class="md-ellipsis">
      Different sequencing platforms
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiplexing" class="md-nav__link">
    <span class="md-ellipsis">
      Multiplexing
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#differential-gene-expression" class="md-nav__link">
    <span class="md-ellipsis">
      Differential gene expression
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#experimental-design" class="md-nav__link">
    <span class="md-ellipsis">
      Experimental Design
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Experimental Design">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#replicates" class="md-nav__link">
    <span class="md-ellipsis">
      Replicates
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#confounding" class="md-nav__link">
    <span class="md-ellipsis">
      Confounding
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#batch-effects" class="md-nav__link">
    <span class="md-ellipsis">
      Batch effects
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Batch effects">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-to-know-whether-you-have-batches" class="md-nav__link">
    <span class="md-ellipsis">
      How to know whether you have batches?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#best-practices-regarding-batches" class="md-nav__link">
    <span class="md-ellipsis">
      Best practices regarding batches:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      Citation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>07 week4 intro to RNAseq</h1>

<div><h2 id="learning-objectives">Learning Objectives:</h2>
<ul>
<li>Understand applications of RNA sequencing</li>
<li>Introduce the overall differential expression workflow</li>
<li>Understand experimental design concepts such as replicates and batch effects </li>
<li>Understand different types of library preps, their requirements and uses. </li>
</ul>
<h2 id="overview-of-rna-seq">Overview of RNA-seq</h2>
<p>RNA-seq is an exciting experimental technique that is utilized to explore and/or quantify gene expression within or between conditions. </p>
<p>As we know, genes provide instructions to make proteins, which perform some function within the cell. Although <strong>all cells contain the same DNA sequence</strong>, muscle cells are different from nerve cells and other types of cells because of the <strong>different genes that are turned on in these cells and the different RNAs and proteins produced</strong>. </p>
<figure>
  <img alt="Gene Expression in Cells" src="../../img/gene_expression_cells.png" width="600">
</figure>
<p>Different biological processes, as well as mutations, can affect which genes are turned on and which are turned off, in addition to, <em>how much</em> specific genes are turned on/off.</p>
<p>To make proteins, the DNA is transcribed into messenger RNA, or mRNA, which is translated by the ribosome into protein. However, some genes encode RNA that does not get translated into protein; these RNAs are called non-coding RNAs, or ncRNAs. Often these RNAs have a function in and of themselves and include rRNAs, tRNAs, and siRNAs, among others. All RNAs transcribed from genes are called transcripts.</p>
<figure>
  <img alt="Gene Products" src="../../img/Gene_products.png" width="400">
</figure>
<p>To be translated into proteins, the RNA must undergo processing to generate the mRNA. In the figure below, the top strand in the image represents a gene in the DNA, comprised of the untranslated regions (UTRs) and the open read frame. Genes are transcribed into pre-mRNA, which still contains the intronic sequences.  After post-transciptional processing, a 5' cap and polyA tail are added and the introns are spliced out to yield mature mRNA transcripts, which can be translated into proteins.</p>
<figure>
  <img alt="Gene Structure" src="../../img/Gene_structure.png" width="600">
</figure>
<p><strong>While mRNA transcripts have a polyA tail, many of the non-coding RNA transcripts do not.</strong></p>
<h3 id="transcriptomics">Transcriptomics</h3>
<p>The transcriptome is defined as a collection of all the transcript readouts present in a cell. RNA-seq data can be used to explore and/or quantify the transcriptome of an organism, which can be utilized for the following types of experiments:</p>
<ul>
<li><strong>Differential Gene Expression</strong>: <em>quantitative</em> evaluation and comparison of transcript levels between conditions</li>
<li><strong>Transcriptome assembly</strong>: building the profile of transcribed regions of the genome, a <em>qualitative</em> evaluation</li>
<li><strong>Refinement of gene models</strong>: building better gene models and verifying them using transcriptome assembly</li>
<li><strong>Metatranscriptomics</strong>: community transcriptome analysis</li>
</ul>
<h2 id="illumina-sequencing">Illumina Sequencing</h2>
<h3 id="illumina-library-preparation">Illumina Library preparation</h3>
<p>The general workflow for library preparation is detailed in the step-by-step images below.</p>
<p><strong>Briefly, the RNA is isolated from the sample and contaminating DNA is removed with DNase.</strong></p>
<figure>
  <img alt="Library Prep" src="../../img/libraryprep_step1-2.png" width="400">
</figure>
<p><strong>The RNA sample then undergoes either selection of the mRNA (polyA selection) or depletion of the rRNA. The resulting RNA is fragmented.</strong> </p>
<blockquote>
<p>Generally, ribosomal RNA represents the majority of the RNAs present in a cell, while messenger RNAs represent a small percentage of total RNA, ~2% in humans. Therefore, if we want to study the protein-coding genes, we need to enrich for mRNA or deplete the rRNA. </p>
</blockquote>
<figure>
  <img alt="PolyA Tail" src="../../img/polyA.png" width="450">
</figure>
<blockquote>
<p><em>The size of the target fragments in the final library is a key parameter for library construction. DNA fragmentation is typically done by physical methods (i.e., acoustic shearing and sonication) or enzymatic methods (i.e., non-specific endonuclease cocktails and transposase tagmentation reactions.</em></p>
</blockquote>
<figure>
  <img alt="Library Prep" src="../../img/libraryprep_step3.png" width="450">
</figure>
<p><strong>The RNA is then reverse transcribed into double-stranded cDNA and sequence adapters are then added to the ends of the fragments.</strong></p>
<blockquote>
<p>The cDNA libraries can be generated in a way to retain information about which strand of DNA the RNA was transcribed from. Libraries that retain this information are called stranded libraries, which are now standard with Illumina’s TruSeq stranded RNA-Seq kits. Stranded libraries should not be any more expensive than unstranded, so there is not really any reason not to acquire this additional information. </p>
<p>There are 3 types of cDNA libraries available:</p>
<ul>
<li>Forward (secondstrand) – reads resemble the gene sequence or the secondstrand cDNA sequence</li>
<li>Reverse (firststrand) – reads resemble the complement of the gene sequence or firststrand cDNA sequence (TruSeq)</li>
<li>Unstranded</li>
</ul>
</blockquote>
<figure>
  <img alt="Library Prep Continued" src="../../img/libraryprep_step4-5.png" width="450">
</figure>
<p><strong>Finally, the fragments are PCR amplified if needed, and the fragments are size selected (usually ~300-500bp) to finish the library.</strong></p>
<figure>
  <img alt="Library Prep Final Step" src="../../img/libraryprep_step6.png" width="450">
</figure>
<p><em>Image credit: <a href="https://www.nature.com/articles/nrg3068">Martin J.A. and Wang Z., Nat. Rev. Genet. (2011) 12:671–682</a></em></p>
<h3 id="strandedness">Strandedness</h3>
<p>The implication of <strong>stranded</strong> libraries is that one could distinguish whether the reads are derived from the forward or reverse-encoded transcripts. </p>
<figure>
  <img alt="Strandedness" src="../../img/strandedness.png" width="800">
</figure>
<ul>
<li>Red = positive strand </li>
<li>Blue = negative strand </li>
</ul>
<h3 id="single-end-versus-paired-end">Single-end versus Paired-end</h3>
<p>After preparation of the libraries, sequencing can be performed to generate the nucleotide sequences of the ends of the fragments, which are called <strong>reads</strong>. You will have the choice of sequencing a single end of the cDNA fragments (single-end reads) or both ends of the fragments (paired-end reads).</p>
<figure>
  <img alt="Paired End Reads" src="../../img/paired_end_reads.png" width="500">
</figure>
<ul>
<li>SE - Single end dataset =&gt; Only Read1</li>
<li>PE - Paired-end dataset =&gt; Read1 + Read2<ul>
<li>often are 2 separate FastQ files! </li>
</ul>
</li>
</ul>
<p>Generally single-end sequencing is sufficient unless it is expected that the reads will match multiple locations on the genome (e.g. organisms with many paralogous genes), assemblies are being performed, or for splice isoform differentiation. </p>
<h3 id="different-sequencing-platforms">Different sequencing platforms</h3>
<p>There are a variety of Illumina platforms to choose from to sequence the cDNA libraries.</p>
<figure>
  <img alt="Illumina Platforms" src="../../img/illumina_platforms.png" width="600">
</figure>
<p><em>Image credit: Adapted from Illumina</em></p>
<p>Differences in platform can alter the length of reads generated, the quality of reads, as well as the total number of reads sequenced per run and the amount of time required to sequence the libraries. The different platforms each use a different flow cell, which is a glass surface coated with an arrangement of paired oligos that are complementary to the adapters added to your template molecules. <strong>The flow cell is where the sequencing reactions take place</strong>.</p>
<figure>
  <img alt="Flow Cell" src="../../img/flow_cell_oligos.png" width="600">
</figure>
<p><em>Image credit: Adapted from Illumina</em></p>
<h3 id="multiplexing">Multiplexing</h3>
<p>Depending on the Illumina platform (MiSeq, HiSeq, NextSeq), the number of lanes per flow cell, and the number of reads that can be obtained per lane varies widely. <strong>The researcher will need to decide on how many reads they would like per sample</strong> (i.e. the sequencning depth) and then based on the platform you choose calculate how many total lanes you will require for your set of samples. </p>
<p>Typically, charges for sequencing are per lane of the flow cell and you will be able to run multiple samples per lane. Illumina has therefore devised a nice multiplexing method which allows libraries from several samples to be pooled and sequenced simultaneously in the same lane of a flow cell. This method requires <strong>the addition of indices</strong> (within the Illumina adapter) or special barcodes (outside the Illumina adapter) as described in the schematic below.</p>
<figure>
  <img alt="Demultiplexing" src="../../img/demultiplexing.png" width="800">
</figure>
<ul>
<li>
<p><strong>General gene-level differential expression:</strong></p>
</li>
<li>
<p>ENCODE guidelines suggest 30 million SE reads per sample (stranded).</p>
</li>
<li>
<p>15 million reads per sample is often sufficient, if there are a good number of replicates (&gt;3). </p>
</li>
<li>
<p>Use of an HiSeq or NextSeq, or NovaSeq for sequencing </p>
</li>
</ul>
<h2 id="differential-gene-expression">Differential gene expression</h2>
<p>Differential gene expression analysis allows us to explore the gene expression changes that occur in disease or between different conditions, by measuring the quantity of RNA expressed by all genes in each of the different conditions. </p>
<figure>
  <img alt="Differential Gene Expression" src="../../img/diff.png" width="600">
</figure>
<p>Using this analysis we can answer questions such as:</p>
<ul>
<li>
<p>What genes are differentially expressed between conditions?</p>
</li>
<li>
<p>Are there any trends in gene expression over time or across conditions?</p>
</li>
</ul>
<figure>
  <img alt="Trends DE" src="../../img/de-trends.png" width="600">
</figure>
<p><em>Citation: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0138236</em> </p>
<ul>
<li>
<p>Which groups of genes change similarly over time or across conditions?</p>
</li>
<li>
<p>What processes or pathways are important for my condition of interest?</p>
</li>
</ul>
<figure>
  <img alt="Pathways DE" src="../../img/de-path.jpg" width="600">
</figure>
<p><em>Citation: https://elifesciences.org/articles/63003</em> </p>
<p>To perform differential gene expression analysis, we perform the following steps:</p>
<figure>
  <img alt="DE Workflow" src="../../img/de_workflow2019.png" width="300">
</figure>
<h2 id="experimental-design">Experimental Design</h2>
<p>Understanding the steps in the experimental process of RNA extraction and preparation of RNA-Seq libraries is helpful for designing an RNA-Seq experiment and important to consider when selecting a dataset to analyze. There are special considerations that should be highlighted which can greatly affect the quality of a differential expression analysis. </p>
<p>These important considerations include:</p>
<ol>
<li>Number and type of <strong>replicates</strong></li>
<li>Avoiding <strong>confounding</strong> variables </li>
<li>Addressing <strong>batch effects</strong></li>
</ol>
<h3 id="replicates">Replicates</h3>
<p>Experimental replicates can be performed as <strong>technical replicates</strong> or <strong>biological replicates</strong>. </p>
<figure>
  <img alt="Biological Replicates" src="../../img/replicates.png" width="350">
</figure>
<p><em>Image credit: <a href="https://dx.doi.org/10.15252%2Fembj.201592958">Klaus B., EMBO J (2015) <strong>34</strong>: 2727-2730</a></em></p>
<ul>
<li>
<p><strong>Technical replicates:</strong> use the same biological sample to repeat the technical or experimental steps in order to accurately measure technical variation and remove it during analysis. </p>
</li>
<li>
<p><strong>Biological replicates</strong> use different biological samples of the same condition to measure the biological variation between samples. </p>
</li>
</ul>
<p><strong>Biological replicates are absolutely essential for differential expression analysis</strong>. In fact, the more biological replicates, the better for estimates of biological variation and the more precise our estimates of the mean expression levels. This leads to more accurate modeling of our data and identification of more differentially expressed genes.</p>
<figure>
  <img alt="DE Replicates" src="../../img/de_replicates_img.png" width="500">
</figure>
<p><em>Image credit: <a href="https://doi.org/10.1093/bioinformatics/btt688">Liu, Y., et al., Bioinformatics (2014) <strong>30</strong>(3): 301–304</a></em></p>
<p>As the figure above illustrates, <strong>biological replicates are of greater importance than sequencing depth</strong>. The figure shows the relationship between sequencing depth and number of replicates on the number of differentially expressed genes identified [<a href="https://academic.oup.com/bioinformatics/article/30/3/301/228651/RNA-seq-differential-expression-studies-more">1</a>]. Note that an <strong>increase in the number of replicates tends to return more DE genes than increasing the sequencing depth</strong>. Therefore, generally more replicates are better than higher sequencing depth, with the caveat that higher depth is required for detection of lowly expressed DE genes and for performing isoform-level differential expression. </p>
<p>Replicates are almost always preferred to greater sequencing depth for bulk RNA-Seq. </p>
<h2 id="confounding">Confounding</h2>
<p>A confounded RNA-Seq experiment is one where you <strong>cannot distinguish the separate effects of two different sources of variation</strong> in the data. </p>
<p>For example, we know that sex has large effects on gene expression, and if all of our <em>control</em> mice were female and all of the <em>treatment</em> mice were male, then our treatment effect would be confounded by sex. <strong>We could not differentiate the effect of treatment from the effect of sex.</strong></p>
<figure>
  <img alt="Confounding Variables" src="../../img/confounded_design.png" width="500">
</figure>
<p><strong>To AVOID confounding:</strong></p>
<ul>
<li>
<p>Ensure animals in each condition are all the <strong>same sex, age, litter, and batch</strong>, if possible.</p>
</li>
<li>
<p>If not possible, then ensure to split the animals equally between conditions</p>
</li>
</ul>
<figure>
  <img alt="Non Confounded Design" src="../../img/non_confounded_design.png" width="400">
</figure>
<h2 id="batch-effects">Batch effects</h2>
<p>Batch effects are a significant issue for RNA-Seq analyses, since you can see significant differences in expression due solely to the batch effect.</p>
<figure>
  <img alt="Non Confounded Design" src="../../img/batch_effect_pca.png" width="600">
</figure>
<p><em>Image credit: <a href="https://www.biorxiv.org/content/early/2015/08/25/025528">Hicks SC, et al., bioRxiv (2015)</a></em></p>
<h3 id="how-to-know-whether-you-have-batches">How to know whether you have batches?</h3>
<ul>
<li>
<p>Were all RNA isolations performed on the same day and with the same kit?</p>
</li>
<li>
<p>Were all library preparations performed on the same day?</p>
</li>
<li>
<p>Did the same person perform the RNA isolation/library preparation for all samples?</p>
</li>
<li>
<p>Did you use the same reagents for all samples?</p>
</li>
<li>
<p>Did you perform the RNA isolation/library preparation in the same location?</p>
</li>
</ul>
<p>If <em>any</em> of the answers is <strong>‘No’</strong>, then you have batches.</p>
<h3 id="best-practices-regarding-batches">Best practices regarding batches:</h3>
<ul>
<li>
<p>Design the experiment in a way to <strong>avoid batches</strong>, if possible.</p>
</li>
<li>
<p>If unable to avoid batches:</p>
</li>
<li>
<p><strong>Do NOT confound</strong> your experiment by batch:</p>
<p><figure markdown="span">
  <img alt="Confounded Batch" src="../../img/confounded_batch.png" width="300">
</figure></p>
<p><em>Image credit: <a href="https://www.biorxiv.org/content/early/2015/08/25/025528">Hicks SC, et al., bioRxiv (2015)</a></em></p>
</li>
<li>
<p><strong>DO</strong> split replicates of the different sample groups across batches. The more replicates the better!</p>
<p><figure markdown="span">
    <img alt="Batch Effect" src="../../img/batch_effect.png" width="300">
  </figure></p>
<p><em>Image credit: <a href="https://www.biorxiv.org/content/early/2015/08/25/025528">Hicks SC, et al., bioRxiv (2015)</a></em></p>
</li>
<li>
<p><strong>DO</strong> include batch information in your <strong>experimental metadata</strong>. During the analysis, we can regress out the variation due to batch so it doesn’t affect our results if we have that information.</p>
<p><figure markdown="span">
    <img alt="Metadata" src="../../img/metadata_batch.png" width="300">
  </figure></p>
</li>
</ul>
<hr>
<h2 id="citation">Citation</h2>
<p><em>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</em><br>
<em>Authors: Mary Piper, Meeta Mistry, Radhika Khetani</em><br>
<em>Other sources - https://umich-brcf-bioinf.github.io/rnaseq_demystified_workshop/site/Module3a_Design_Prep_Seq#2_Experimental_Design_and_Practicalities</em> </p></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["content.code.copy", "content.code.annotate", "content.code.select", "navigation.tabs", "navigation.top"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.83f73b43.min.js"></script>
      
    
  </body>
</html>